libp2p/stream/bufferstream.nim

```diff
@@ -114,199 +98,89 @@ method initStream*(s: BufferStream) =
-  s.readReqs = initDeque[Future[void]]()
-  s.dataReadEvent = newAsyncEvent()
-  s.lock = newAsyncLock()
-  s.writeLock = newAsyncLock()
-
-  if not(isNil(handler)):
-    s.writeHandler = proc (data: seq[byte]) {.async, gcsafe.} =
-      defer:
-        s.writeLock.release()
-
-      # Using a lock here to guarantee
-      # proper write ordering. This is
-      # specially important when
-      # implementing half-closed in mplex
-      # or other functionality that requires
-      # strict message ordering
-      await s.writeLock.acquire()
-      await handler(data)
+  s.initStream()
 
   trace "created bufferstream", oid = $s.oid
 
-proc newBufferStream*(handler: WriteHandler = nil,
-                      size: int = DefaultBufferSize,
-                      timeout: Duration = DefaultConnectionTimeout): BufferStream =
+proc newBufferStream*(timeout: Duration = DefaultConnectionTimeout): BufferStream =
   new result
   result.timeout = timeout
-  result.initBufferStream(handler, size)
-
-proc popFirst*(s: BufferStream): byte =
-  result = s.readBuf.popFirst()
-  s.dataReadEvent.fire()
-
-proc popLast*(s: BufferStream): byte =
-  result = s.readBuf.popLast()
-  s.dataReadEvent.fire()
+  result.initBufferStream()
 
-proc shrink(s: BufferStream, fromFirst = 0, fromLast = 0) =
-  s.readBuf.shrink(fromFirst, fromLast)
-  s.dataReadEvent.fire()
-
-proc len*(s: BufferStream): int = s.readBuf.len
-
-method pushTo*(s: BufferStream, data: seq[byte]) {.base, async.} =
+proc pushTo*(s: BufferStream, data: seq[byte]) {.async.} =
   ## Write bytes to internal read buffer, use this to fill up the
   ## buffer with data.
   ##
-  ## This method is async and will wait until  all data has been
+  ## This method is async and will wait until all data has been
   ## written to the internal buffer; this is done so that backpressure
   ## is preserved.
   ##
 
-  if s.atEof:
+  if s.isClosed:
     raise newLPStreamEOFError()
 
-  defer:
-    # trace "ended", size = s.len
-    s.lock.release()
-
-  await s.lock.acquire()
-  var index = 0
-  while not s.closed():
-    while index < data.len and s.readBuf.len < s.maxSize:
-      s.readBuf.addLast(data[index])
-      inc(index)
-    # trace "pushTo()", msg = "added " & $s.len & " bytes to readBuf", oid = s.oid
-
-    # resolve the next queued read request
-    if s.readReqs.len > 0:
-      s.readReqs.popFirst().complete()
-      # trace "pushTo(): completed a readReqs future", oid = s.oid
-
-    if index >= data.len:
-      return
-
-    # if we couldn't transfer all the data to the
-    # internal buf wait on a read event
-    await s.dataReadEvent.wait()
-    s.dataReadEvent.clear()
-
-proc drainBuffer*(s: BufferStream) {.async.} =
-  ## wait for all data in the buffer to be consumed
-  ##
+  if data.len == 0:
+    return
 
-  trace "draining buffer", len = s.len
-  while s.len > 0:
-    await s.dataReadEvent.wait()
-    s.dataReadEvent.clear()
+  await s.readQueue.addLast(data)
```

Comment:
What happens if I push a GBs of data here? There aren't any limits on the buffer size.

Reply:
currently, mplex loads data into a seq that it holds in memory, then pushes that seq byte by byte to bufferstream, synchronizing for every byte.

in this PR, the bufferstream has a bounded 1-item queue, so mplex will hold the same seq in memory as it had before (which is limited), then push it in one go to the queue and will not read the next one until this one has been processed.